Open Humans integration and data access work

beaugunderson opened the pull request:
Just creating a WIP pull request for Open Humans integration work that will happen this week.

- [x] add quick-start instructions to `README.md`
- [x] comment out and note that insertAGMultiple is vulnerable to SQL injection (it's currently unused)
- [x] consolidate all data access into `amgut.connections`
- [x] lazily connect to databases (obviates need to disconnect the automatically-connected connections in `scripts/ag`, for example)
- [x] increase robustness of `scripts/ag`
- [x] make sure everything is created in the `ag` schema (including the stored procedures and the `settings` table)
- [x] always specify SEARCH_PATH as "ag, public"
- [x] add a Procfile for ease of running with a specified port
- [x] add labels for login/password field on the main page
- [x] resize the password field to fit a bcrypt hash
- [x] store and verify passwords as bcrypt hashes
- [x] add the Open Humans access token
- [x] investigate an alternate approach for lazy-loading database connections (use `__getattr__` to connect on first attribute access)
- [x] add a command to `scripts/ag` to migrate unencrypted passwords
- [x] add a command or database patch to migrate settings to the `ag` schema
- [x] add a command or database patch to migrate stored procedures to the `ag` schema

### Migration steps

- `scripts/ag patch` (this is idempotent)
- `scripts/ag migrate_passwords` (this is idempotent)

wasade commented on amgut/connections.py:
```diff
@@ -0,0 +1,50 @@
+# connections to the database since it handles them itself.
+
+
+class _LazyAgData(object):
```
This is a very clean way to handle these connections. @josenavas @squirrelo, just pinging the two of you as it might be nice to adopt something like this in qiita

wasade commented on amgut/db/populate_test.sql:
```diff
@@ -1,6 +1,83 @@
+          ('test_ha', '1234', '000000003', '5678', '3', 'n'),
+          ('test_ha', '1234', '000000004', '5678', '3', 'n');
+insert into ag_participant_exceptions (ag_login_id, participant_name)
+    values ('d8592c74-7da1-2135-e040-8a80115d6401', 'execption');
```
minor, but typo: execption -> exception

wasade commented on amgut/lib/data_access/ag_data_access.py:
```diff
@@ -189,6 +197,7 @@ def getAGLogins(self):
 
         return return_res
 
+    # note: only used by the password migration
```
?

wasade replied:
I see now, ignore the comment

wasade commented on amgut/lib/data_access/ag_data_access.py:
```diff
@@ -428,14 +441,16 @@ def getConsent(self, survey_id):
-                                        field_name, field_value)
-        self.get_cursor().execute(sql)
-        self.connection.commit()
+    # NOTE: This method is vulnerable to SQL injection as-written but is not
```
Much of the data access code is carryover from the old mod_python + Oracle deployment, and there are blocks of code that slipped through that should be removed. If this is not in use then please delete it.

beaugunderson replied:
The one place where it was in use is inside the animal survey, but the bulk of that survey is commented out (including the line that references this function); I figured it may be used again at some point from there but wanted to note that potential for SQL injection so that it could be fixed before it was used again. Should it just be removed?

wasade replied:
Yes please, thanks!

Flagging @adamrp just to make sure he saw this note re: animal survey. We
haven't redone it yet, so it's still carrying some fluff.

On Fri, Jan 2, 2015 at 8:22 PM, Beau Gunderson <notifications@github.com>
wrote:

> In amgut/lib/data_access/ag_data_access.py
> <https://github.com/biocore/american-gut-web/pull/354#discussion-diff-22429153>
> :
>
> > @@ -428,14 +441,16 @@ def getConsent(self, survey_id):
> >              if result:
> >                  return {k: v for k, v in zip(colnames, result)}
> >
> > -    def insertAGMultiple(self, ag_login_id, participant_name, field_name,
> > -                         field_value):
> > -        sql = ("insert into ag_survey_multiples (ag_login_id, "
> > -               "participant_name,item_name, item_value) values ('{0}','{1}',"
> > -               " '{2}', '{3}')").format(ag_login_id, participant_name,
> > -                                        field_name, field_value)
> > -        self.get_cursor().execute(sql)
> > -        self.connection.commit()
> > +    # NOTE: This method is vulnerable to SQL injection as-written but is not
>
> The one place where it was in use is inside the animal survey, but the
> bulk of that survey is commented out (including the line that references
> this function); I figured it may be used again at some point from there but
> wanted to note that potential for SQL injection so that it could be fixed
> before it was used again. Should it just be removed?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/biocore/american-gut-web/pull/354/files#r22429153>.
>

wasade commented on amgut/lib/util.py:
```diff
@@ -169,8 +170,9 @@ def survey_vioscreen(survey_id):
+    user_info = ag_data.get_person_info(survey_id)
     url = ("https://vioscreen.com/remotelogin.aspx?Key=%s&RegCode=KLUCB" %
-           url_escape(encrypt_key(survey_id)))
+           url_escape(encrypt_key(survey_id, user_info)))
```
@beaugunderson, can you comment as to why this was done? @squirrelo, pinging you to keep you in the loop since you're most familiar with this interaction.

wasade replied:
...should have simply scrolled a little further. Makes sense now, thanks. @squirrelo, it would still be useful if you could once over these particular changes though

squirrelo replied:
Not sure why the change to getting user info here, but works for me. It doesn't look like it changes any fundamental code.

beaugunderson replied:
Just consolidating things that access the database in one place, was purely organizational.l :)

squirrelo replied:
Cool :+1:

wasade commented on scripts/ag:
```diff
@@ -105,32 +147,118 @@ def clean_test():
+        click.echo('Revoking all privileges on all functions in ag')
+        cur.execute('REVOKE ALL PRIVILEGES ON ALL FUNCTIONS IN SCHEMA ag '
+                    'FROM ag_wwwuser CASCADE')
+    except:
```
If one of the `execute`s excepts then the rest won't execute. If the intent is to ensure all these are executed regardless of if anything excepts (which seems to be the case given the `pass`) then this structure may need to be revisited. However, this does not hamper production and only test environments, so its exposure is limited.

beaugunderson replied:
True, probably better to make these into functions and loop through them, `pass`ing on each--but seems OK to leave for now.

wasade commented on scripts/ag:
```diff
@@ -105,32 +147,118 @@ def clean_test():
+    click.echo('Dropping database {}'.format(AMGUT_CONFIG.database))
 
-@env.command()
+    command = ['dropdb', '--if-exists']
```
Why fork instead of issuing via psycopg2? Just generally curious

beaugunderson replied:
This was done early on as I was still learning how the automatic connection code worked; I thought a connection was being held open (and that was preventing the database from being dropped). After I refactored to use the lazy-loaded connections that issue went away, but I think this did fix another issue in my system:

I setup my database so that the amgut user only had access to the amgut database. To delete the database via psycopg2 I don't think you can specify the database to connect to (since you'll then be prevented from dropping it). There are no other databases my user had access to from which to delete the database, but deleting with dropdb did work.

wasade replied:
Ah, good point, that makes sense. Thanks!

On Fri, Jan 2, 2015 at 8:27 PM, Beau Gunderson <notifications@github.com>
wrote:

> In scripts/ag
> <https://github.com/biocore/american-gut-web/pull/354#discussion-diff-22429169>
> :
>
> >
> > -@env.command()
> > +    command = ['dropdb', '--if-exists']
>
> This was done early on as I was still learning how the automatic
> connection code worked; I thought a connection was being held open (and
> that was preventing the database from being dropped). After I refactored to
> use the lazy-loaded connections that issue went away, but I think this did
> fix another issue in my system:
>
> I setup my database so that the amgut user only had access to the amgut
> database. To delete the database via psycopg2 I don't think you can specify
> the database to connect to (since you'll then be prevented from dropping
> it). There are no other databases my user had access to from which to
> delete the database, but deleting with dropdb did work.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/biocore/american-gut-web/pull/354/files#r22429169>.
>

wasade commented:
Done with static analysis. This looks very good, thank you for the contributions, only a few minor comments. I'll pull this down either over the weekend or first thing Monday to test locally. 

@adamrp @ElDeveloper @squirrelo @josenavas, is anyone else able to pull down an run locally as well? I'd like to get this merged into master by mid next week at the latest, as well as any other pending PRs that need to get in.

beaugunderson commented:
@wasade OK, I think I addressed all your comments :)

wasade commented:
@beaugunderson, just pulled down to test local. First thing that popped up:

```bash
./ag drop_test
/Users/mcdonadt/rs/american-gut-web/amgut/lib/config_manager.py:33: UserWarning: Extra main section option(s) found: 'debug'
  warnings.warn("Extra %s found: %r" % (set_type, extra))
Dropping user ag_wwwuser
Reassigning owned objects
Revoking all privileges on ag2
Revoking all privileges in schema public
Revoking all privileges on all tables in public
Revoking all privileges on all functions in public
Revoking all privileges in schema ag
Revoking all privileges on all tables in ag
Revoking all privileges on all functions in ag
Dropping user ag_wwwuser...
Dropping schema ag...
Dropping database ag2
dropdb: unrecognized option `--if-exists'
Try "dropdb --help" for more information.
$ dropdb --version
dropdb (PostgreSQL) 9.1.9
```

beaugunderson commented:
@wasade Ah, looks like `--if-exists` was added in postgresql 9.2; does american-gut-web target 9.1? (Worth updating the README if so, which I can do)

wasade commented:
I reset one of the `kit_password`s in the test database, and the performed the password migration. I was unable to login using that account after, indicating an invalid password. Below are the steps performed. I tried to login with the plain text which threw a 500 error (that's okay I think for now as this should never happen in practice). I also tried logging in following the migration and was met with an invalid kit ID or password.

```bash
$ echo "update ag.ag_kit set kit_password='foo' where supplied_kit_id='test';" | psql -d ag2
Expanded display is used automatically.

UPDATE 1
Time: 5.102 ms
$ echo "select kit_password from ag.ag_kit where supplied_kit_id='test';" | psql -d ag2
Expanded display is used automatically.

+--------------+
| kit_password |
+--------------+
| foo          |
+--------------+
(1 row)

Time: 2.452 ms
$ ./ag migrate_passwords
Connecting to postgres for amgut.connections.ag_data
kit 1111 is already a bcrypt hash
migrating kit test to a bcrypt hash
$ echo "select kit_password from ag.ag_kit where supplied_kit_id='test';" | psql -d ag2
Expanded display is used automatically.

+--------------------------------------------------------------+
|                         kit_password                         |
+--------------------------------------------------------------+
| $2a$12$IRfEN4AAkOjv7ck9ErGlYe5mqQphHwLx.lvisIrIStjEVFcSTmxQS |
+--------------------------------------------------------------+
(1 row)

Time: 2.310 ms
```

wasade commented:
Those were the only issues I encountered! re: version, that's a good question. On production, we're actually using 9.3.5, so my env is actually a little behind. I think it is okay to set the minimum to 9.2 (eg >= 9.2). Sound good?

beaugunderson commented:
@wasade Sounds good re: 9.2; and I'll investigate the password reset issue!

wasade commented:
Great, thanks! I'll be in the air shortly, but will re-review around 11:30
PT if you're able to get commits in before then.

On Mon, Jan 5, 2015 at 12:49 PM, Beau Gunderson <notifications@github.com>
wrote:

> @wasade <https://github.com/wasade> Sounds good re: 9.2; and I'll
> investigate the password reset issue!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/biocore/american-gut-web/pull/354#issuecomment-68765318>
> .
>

beaugunderson commented:
@wasade OK, the password issue is fixed. I implemented migration before I added encryption to the rest of the data access functions; once I did the migration script was actually hashing the passwords twice. :)

wasade commented:
Great, thanks!
